realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -66,6 +70,61 @@
     Handler handler;
     HandlerController handlerController;
 
+    // Map between RealmConfig and SharedGroup to stop a Realm from waiting
+    // Only one thread at a time can operation on Collections.synchronizedMap
+    protected static final Map<String, WeakReference<? extends BaseRealm>> waitingRealms =
+        Collections.synchronizedMap(new HashMap<String, WeakReference<? extends BaseRealm>>());
+
+    protected static <R extends BaseRealm> void addRealmToWaitList(R realm) {
+        if (realm == null) {
+            return;
+        }
+        waitingRealms.put(realm.getPath(), new WeakReference<R>(realm));
+    }
+
+    protected static <R extends BaseRealm> void removeRealmFromWaitList(R realm) {
+        if (realm == null) {
+            return;
+        }
+        String realmPath = realm.getPath();
+        Iterator<Map.Entry<String, WeakReference<? extends BaseRealm>>> iterator = waitingRealms.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<String, WeakReference<? extends BaseRealm>> entry = iterator.next();
+            String waitingRealmPath = entry.getKey();
+            WeakReference<? extends BaseRealm> weakRealm = entry.getValue();
+            if (weakRealm.get() == null || (realmPath.equals(waitingRealmPath) && weakRealm.get() == realm)) {
+                iterator.remove();
+            }
+        }
+    }
+
+    /**
+     * Makes all Realm in shared database which called {@link #waitForChange()} return {@code false}
+     * immediately. This method is safe to be called in any thread.
+     *
+     * @param realm to find other waiting Realm in shared database.
+     * @throws IllegalArgumentException if realm is null.
+     */
+    public static <R extends BaseRealm> void stopAllWaitingRealmInSharedDatabase(R realm) {
```

Comment:
This seems like overkill? Especially considering all the caching it takes with it and that it might have unwanted side-effects, like having two threads waiting and you only want to release 1?

What not just let `stopWaitForChange()` be a thread safe method on a realm instance?
Or even better, this should just work if you call `thread.interrupt()` as the the default way of interrupting a thread that is in a waiting state() ?

Thoughts @realm/java?